Lending Market v1 – Smart Contract Architecture & Implementation Review

Enricrypto opened the pull request:
This is the first complete version of a decentralized lending market protocol implemented in Solidity. The system allows users to:

- Deposit collateral

- Borrow against collateralized positions

- Accrue interest dynamically based on utilization

- Be liquidated if their health factor falls below a safe threshold

The protocol architecture includes:

1. - Vault Contract – Manages deposited liquidity and tracks total assets

2. - Market Contract – Handles borrow/lend logic, interest indexing, and liquidation triggers

3. - Interest Rate Model – Implements a jump-rate model with adjustable base rate, slopes, and utilization thresholds
 
4. - Price Oracle Integration – Fetches real-time token prices for accurate collateral/debt valuation
 
5. - Liquidation Mechanics – Supports partial or full liquidation with penalty and protocol fee distribution
 
6. - Protocol Fee Management – Configurable treasury fee on interest accrued

Enricrypto committed:
first code review

Enricrypto committed:
fix: test issues

Enricrypto committed:
fix: test market.t.sol - depositLendToken

Enricrypto committed:
feat: add function registerLendDeposit

Enricrypto committed:
fix: depositLendToken and withdrawLendToken

Enricrypto committed:
fix: tests for depositLendToken, withdrawLendToken and borrow

Enricrypto committed:
feat: interestRateModel contract

Enricrypto committed:
forge install: chainlink-local

7d8b2f888e1f10c8841ccd9e0f4af0f5baf11dab

Enricrypto committed:
feat: interest rate model contract

Enricrypto committed:
feat: market integration with price oracle and interest rate model contracts

Enricrypto committed:
feat: removeCollateralToken

Enricrypto committed:
feat: add function getLendingRate - Interest Rate Model contract

Enricrypto committed:
fix: logic relation vault and market contracts + interaction user with loan assets and vault

Enricrypto committed:
delete: tracking for lenders on loan assets

Enricrypto committed:
fix: modify function getSlope - different asset risk profiles

Enricrypto committed:
fix: updatePriceVolatility function

Enricrypto committed:
feat: added calculateRepaymentToVault and repayToVault functions

Enricrypto committed:
fix: fixed repayTovaul function and repay function to track total repaid by borrowers

Enricrypto committed:
fix: function repay, update borrowerPrincipal tracking after repayment

Enricrypto committed:
fix: issues with vault contract

Enricrypto committed:
fix: vault contract issues - using asset from ERC4626, fix total assets function, fix modifier

Enricrypto committed:
fix: vault contract - withdraw function arguments

Enricrypto committed:
fix: vault contract - totalAssets function

Enricrypto committed:
fix: interest rate model contract - implemented jump rate model

Enricrypto committed:
feat: newMarket: implementing logic of market in this contract

Enricrypto committed:
fix: getTotalSupply and getTotalBorrows functions - Interest rate Model contract

Enricrypto committed:
fix: delete old market contract

Enricrypto committed:
fix: borrow, borrowRateAtTime and borrowerInterestAccrued functions

Enricrypto committed:
feat: updateGlobalBorrowIndex function on market contract, getBorrowRatePerBlock function on InterestRateModel contract, fix borrow logic and borrowerInterestAccrued logic

Enricrypto committed:
feat: getTotralInterestAccrued and borrowedPlusInterest functions - Market contract

Enricrypto committed:
fix: scaling precision when calculating newGlobalBorrowIndex within the updateGlobalBorrowIndex and getTotalInterestAccrued functions

Enricrypto committed:
fix: rearrange calculation of globalBorrowIndex to avoid overflow

Enricrypto committed:
fix: borrow function logic order

Enricrypto committed:
fix: getUserTotalDebt function, added updateGlobalBorrowIndex() ensuring no interest is skipped

Enricrypto committed:
feat: add totalLiquidity function to vault contract, fix: add checkLiquidity avaialable to borrow function to avoid overborrowing

Enricrypto committed:
feat: added reentrancyGuard, checks for transfers, check for maxborrowing function

Enricrypto committed:
fix: logic of borrow function, eliminate duplicated interest accrued

Enricrypto committed:
feat: repay function

Enricrypto committed:
fix: refactor helper functions market contract

Enricrypto committed:
fix: check for collateral token balance before removing collateral from market

Enricrypto committed:
fix: eliminate reserve factor from interest rate model contract

Enricrypto committed:
fix: updater global borrow index on functions, change functions to internal

Enricrypto committed:
fix: totalAssets, adminRepay & adminBorrow functions

Enricrypto committed:
feat: functions previewDeposit, previewWithdraw, maxWithdraw and maxRedeem on vault contract

Enricrypto committed:
fix: add updateGlobalBorrowIndex for depositCollateral and withdrawCollateral functions, remove previewDeposit and previewWithdraw from Vault contract

Enricrypto committed:
fix: totalAssets function, remove zero case logic

Enricrypto committed:
feat: maxWithdraw, maxRedeem, setMarket functions on vault contract - add price feed parameter to addCollateralToken function - tests for vauklt and market contracts

Enricrypto committed:
fix: interest calculated by time, instead of blocks

Enricrypto committed:
fix: logic to calculate global index and update index on klast user's interaction, borrowerInterestAccrued relies on global index rather than elapsed time

Enricrypto committed:
fix: updateInterestGlobalBorrowIndex, borrowerInterestAccrued, borrow and repay functions logic to accrue interests and calculate repayments and user debts

Enricrypto committed:
fix: _updateInterestGlobalBorrowIndex function logic to update index based on utilization rate and dynamic borrow rates

Enricrypto committed:
feat: liquidation logic, including _transferCollateral and _calculateCollateralToLiquidate functions

Enricrypto committed:
fix: getUserTotalCollateralValue function

Enricrypto committed:
fix: health factor function, include liquidation threshold

Enricrypto committed:
fix: borrow function

Enricrypto committed:
fix: liquidate function, logic updateInterestGlobalBorrowIndex, logic borrow and repay

Enricrypto committed:
fix: updateInterestGlobalBorrowIndex logic

Enricrypto committed:
fix: logic update global index

Enricrypto committed:
fix: clean market contract code

Enricrypto committed:
feat: add lending rate function

RedVeil commented on src/Vault.sol:
```diff
@@ -4,38 +4,152 @@ pragma solidity ^0.8.0;
+        _;
+    }
+
+    modifier onlyValidMarket() {
```
you never use this from what i can see. can be removed

RedVeil commented on src/Vault.sol:
```diff
@@ -4,38 +4,152 @@ pragma solidity ^0.8.0;
 
     /// @notice Deposit ERC-20 tokens into the vault
     function deposit(
-        uint256 assets,
```
this can be removed since it essentially works like the normal 4626. If you want the nonReentrant than make it just `return super.deposit(amount, receiver)` the other stuff isnt necessary

RedVeil replied:
BUT if you go for the nonReentrant version you also have to do with for `mint` and `redeem`

RedVeil commented on src/Vault.sol:
```diff
@@ -4,38 +4,152 @@ pragma solidity ^0.8.0;
         return shares;
     }
 
     function withdraw(
```
same comment as for deposit

RedVeil commented on src/Vault.sol:
```diff
@@ -4,38 +4,152 @@ pragma solidity ^0.8.0;
+    // Admin function to borrow tokens, only callable by the market contract
+    function adminBorrow(uint256 amount) external nonReentrant onlyMarket {
+        IERC20 token = IERC20(asset());
+        require(address(token) != address(0), "Asset token is not set");
```
that case doesnt need to be checked since its impossible to create a vault without the token. This also means you dont need to store token. you can simply call the asset.transfer directly down below

RedVeil commented on src/Vault.sol:
```diff
@@ -4,38 +4,152 @@ pragma solidity ^0.8.0;
+    // Admin function to repay tokens back to the vault, only callable by the market contract
+    function adminRepay(uint256 amount) external onlyMarket {
+        IERC20 token = IERC20(asset());
+        require(address(token) != address(0), "Asset token is not set");
```
same comment as for `adminBorrow`

RedVeil commented on src/Vault.sol:
```diff
@@ -4,38 +4,152 @@ pragma solidity ^0.8.0;
+    function setMarket(address _market) external onlyOwner {
+        require(address(market) == address(0), "Market already set");
+        require(_market != address(0), "Invalid market address");
+        market = Market(_market);
```
we should also check that the market loan asset is the same as this vaults asset. Since you cant change the market afterwards this might lead to an issue otherwise where you pass the wrong market address and brick the vault

RedVeil commented on src/Vault.sol:
One thing that is still missing is a second ERC4626 which we send all the idle assets to

RedVeil commented on src/PriceOracle.sol:
```diff
@@ -0,0 +1,68 @@
+        );
+        require(
+            address(priceFeeds[asset]) != address(0),
+            "Feed does not exist"
```
Would be the question if you want to have this or if you want to have the option to swap out price feeds. It would be a security risk since you can add a faulty one. BUT you could also fix or upgrade a price feed in case your current one breaks or gets discontinued

Enricrypto replied:
This is explicitly checking that a feed already exists. I'm enforcing that only existing feeds can be updated. However, I should think if this function doesn't present a security risk, because of a malicious market creator or by the market creator making a mistake and assigning the wrong price feed for a token.

RedVeil commented on src/InterestRateModel.sol:
```diff
@@ -0,0 +1,106 @@
+
+    // Calculate utilization rate: U = totalBorrows / totalAssets
+    function getUtilizationRate() public view returns (uint256) {
+        uint256 totalAssets = getTotalAssets();
```
im wondering if we should only look at idle assets here and not totalAssets since totalAssets would include interest. But since the interest doesnt exist yet i feel like it skews the utilisation a bit. Worth a discussion to think about the consequences of either option

Enricrypto replied:
Since interests isn’t available liquidity, I think it would be better to calculate utilisation rate based on idle assets and total borrows, without including the unrealised interests. Interests wouldn't be a lendable capital, so I think it will be better the way you mentioned.

RedVeil commented on src/InterestRateModel.sol:
```diff
@@ -0,0 +1,106 @@
+        uint256 totalAssets = getTotalAssets();
+        uint256 totalBorrows = getTotalBorrows();
+        if (totalAssets == 0) return 0; // avoid division by zero
+        return (totalBorrows * 1e18) / totalAssets;
```
id use MulDiv whereever possible

RedVeil commented on src/InterestRateModel.sol:
```diff
@@ -0,0 +1,106 @@
+            // Above optimal utilization: use slope2 (steep increase)
+            // The "kink" occurs at the point where the utilization rate exceeds optimalUtilization.
+            uint256 excessUtilization = utilization - optimalUtilization;
+            return
```
thats a very fair solution you came up with here to only apply the slope2 on excess capital. Normally we add it to the whole utilisation and just use slope2. Simplifies logic and makes overall interest more expensive nudging people to replay to get back to ideal utilisation. The goal is to always have optimal utilisation since higher utilisation would prevent lenders from withdrawing. so we always want to have some idle capital. if we dont punish high utilisation enough this might create a scenario where lenders cant withdraw since borrowers still seem it as profitable to borrow at a slightly higher rate

Enricrypto replied:
understood. Fixed.

RedVeil commented on src/InterestRateModel.sol:
```diff
@@ -0,0 +1,106 @@
+
+        if (utilization < optimalUtilization) {
+            // Below optimal utilization: use slope1
+            return baseRate + (utilization * slope1) / 1e18;
```
also use mulDiv here probably

RedVeil commented on src/Market.sol:
```diff
@@ -1,297 +1,863 @@
-    mapping(address => uint256) public ltvRatios; // Token -> LTV ratio (percentage out of 100)
+    uint256 public totalBorrows;
+    uint256 public globalBorrowIndex = 1e18;
+    uint256 public lastAccruedInterest;
```
TBD

RedVeil commented on src/Market.sol:
```diff
@@ -1,297 +1,863 @@
-    // Array to track all supported collateral tokens
-    address[] public collateralTokens;
+    // --- Collateral Management ---
+    address[] public collateralTokenList;
```
TBD

RedVeil commented on src/Market.sol:
```diff
@@ -1,297 +1,863 @@
+        require(!supportedCollateralTokens[token], "Token already added");
+
+        supportedCollateralTokens[token] = true;
+        collateralTokenList.push(token);
```
TBD

RedVeil commented on src/Market.sol:
```diff
@@ -1,297 +1,863 @@
+    function removeCollateralToken(address token) external onlyOwner {
+        require(supportedCollateralTokens[token], "Token not supported");
+        require(depositsPaused[token], "Pause deposits first");
+        require(
```
TBD

Enricrypto committed:
Refactored & Implemented Functions: _validateAndCalculateLiquidation, _isHealthy, _seizeCollateral, _seizeOneCollateral, borrow(Enforces borrow constraint based on LLTV, not health factor)

Enricrypto committed:
refactor: use Math.mulDiv for better precision in getLendingRate and other math ops

Enricrypto committed:
fix: _validateAndCalculateFullLiquidation, allows for full liquidation

Enricrypto committed:
fix: casting and var declaration issues

Enricrypto committed:
fix: issue with setMarket validation

RedVeil commented on src/Market.sol:
```diff
@@ -1,297 +1,800 @@
+    }
+
+    // Returns the total amount of a specific collateral token held by the contract
+    function _getTotalCollateralLocked(
```
you dont really need this function. its only used once

RedVeil commented on src/Market.sol:
```diff
@@ -1,297 +1,800 @@
+    }
+
+    // Calculates a user's total borrowing power (in USD) based on their collateral and the market's max LTV
+    function _getUserTotalBorrowingPower(
```
this makes sense. but simplify it as `_getUserTotalCollateralValue() * lltv`

Enricrypto replied:
got it, i was doing the same logic on _getUserTotalCollateralValue()

Enricrypto committed:
fix: repay logic, market contract

Enricrypto committed:
fix: update liquidation health check and simplify collateral logic

- Updated _isHealthy function to factor in liquidation penalty, enabling
  earlier and more accurate liquidation triggers.
- Removed _getTotalCollateralLocked and _getUserTotalBorrowingPower
  functions as they were only used once, simplifying code and improving clarity.

Enricrypto committed:
fix: paused collateral function

Enricrypto committed:
refactor: remove global collateral list, improve collateral tracking, visibility updates

- Removed collateralTokenList and updated logic to use userCollateralAssets mapping
- Refactored _getUserTotalCollateralValue and _getUserBorrowableCollateralValue
- Renamed and exposed totalBorrowsWithInterest, getUserTotalDebt
- Updated internal/private function visibility and usage
- Cleaned up add/remove collateral token logic

Enricrypto committed:
fix: test call issues

Enricrypto committed:
fix: allow withdrawal of paused collateral if position remains healthy

Enricrypto committed:
feat: add StrategyManager system with IStrategy interface and AaveStrategy implementation

Enricrypto committed:
feat: integrate strategy logic into Vault with totalAssets override and lending support

Enricrypto committed:
fix: change logic _getUserTotalCollateralValue to exclude paused tokens

Enricrypto committed:
fix: onlyMarket modifier for setMarket, fix setup function on testing

Enricrypto committed:
fix: properly account vault totalAssets as sum of strategy assets and lent assets with interest

- Override totalAssets() to return strategy assets + market totalBorrowsWithInterest()
- Add totalStrategyAssets() and availableLiquidity() helper views for better accounting
- Update maxWithdraw() to calculate user withdrawable assets based on strategy assets and share ownership
- Fix lending rate calculation to correctly use totalAssets() instead of shares
- Keeps ERC4626 share accounting intact while reflecting true asset backing

Enricrypto committed:
fix: add reentrancy protection and validations to strategy change logic

- Marked changeStrategy() as nonReentrant to prevent race conditions during strategy transitions.
- Added asset compatibility check to ensure the new strategy matches the Vault's asset.
- Emitted StrategyChanged event to log strategy transitions for better observability and auditing.

Enricrypto committed:
fix: totalSupply variable changed for totalAssets

Enricrypto committed:
fix: normalize units in liquidation and repayment logic

- Use 18-decimal normalization for all internal calculations
- Denormalize values for vault and treasury transfers
- Add helper for loan token decimals
- Fix event emission and early return in _seizeCollateral

Enricrypto committed:
Refactor: reorganize functions and enhance liquidation and bad debt handling

- Clear grouping of functions by category: debt calculations, decimals, user health, and helpers.
- Updated _validateAndCalculateFullLiquidation to include bad debt handling.
- Implemented function to transfer unrecoverable debt to bad debt management account.
- Improved calculations and conversions between normalized amounts and native token decimals.
- Optimized partial and full liquidation flow for better clarity and robustness.

Enricrypto committed:
Improve liquidation logic, tests, and collateral handling

- Renamed _validateAndCalculateFullLiquidation to _validateAndCalculateMaxLiquidation for clearer intent.
- Enhanced liquidation logic to handle cases where collateral value is less than debt, tracking unrecovered bad debt correctly.
- Added comments explaining liquidation calculations, unrecovered debt management, and bad debt transfer.
- Refined test suite by replacing console logs with precise asserts, including detailed comments on each assertion for easier debugging.
- Fixed collateral deposit logic by ensuring proper token approvals and transfers, confirming correct normalization for 18-decimal tokens like WETH.
- Added safety checks and clarified partial vs full liquidation handling.

Enricrypto committed:
refactor(Vault): improve asset flow logic and test coverage for mint, redeem, and admin functions

- Capped  and  based on strategy liquidity
- Auto-deposit into strategy after , and withdraw before  or
- Added comprehensive unit tests for minting, redeeming, borrowing, and repaying
- Simulated ERC20 balances with  to support testing without mint functions
- Strengthened balance assertions to ensure safety and correctness

Enricrypto committed:
Add tests for edge cases and analyze protocol behavior on liquidity and fees

- Implement tests for failure scenarios: withdrawals exceeding available liquidity, minting without sufficient funds, unauthorized admin calls
- Review nonReentrant protection and decide to skip complex reentrancy tests
